### DISCUSSION TITLE
Startup script location

### DISCUSSION THREAD

**Discussion by enyafox** [1000001101000/Python_buffalo_libmicon#3] [Q&A] [answered]
So I have noticed that all the scripts runs on startup. Does anyone know where the file location would be? 
Thanks.

**Comment by 1000001101000** [comment, marked as answer]
Good to hear from you!

I'm not entirely sure I understand the question. Are you asking how they work in my Debian installers?

For those projects I have it setup for them to be installed into `/usr/local/bin/micon_scripts/`. The startup scripts are called by systemd services at startup, the shutdown script is linked in `/lib/systemd/system-shutdown/` so that it gets called at shutdown. These locations are specific to Debian with SystemD.

They Python scripts themselves aren't actually Debian/SystemD specific, you could hypothetically use them with other linux distros or even other Operating Systems if you adjust the device names in the scripts. In that case you'd use that systems startup/shutdown mechanisms to call them.

Let me know if that answers the question.

**Reply by enyafox** [reply]
I was talking about after the installation. The directories are exactly where you have specified. I just want to do some customization to have the NAS stand out and make a small selection like the original firmware had to show different information. Thanks for the help. The installation was a little tricky at first but I got there in the end and its easy to memorize.

**Reply by 1000001101000** [reply]
Which model device are you using?

**Reply by enyafox** [reply]
I'm using the TS-XL. I am only using one drive at the moment. I did wipe the drive on a Debian live usb and did the boot partition manually. It booted, forgot I had to use PUTTY to SSH and that worked. The guided partitioning works, setting up boot partition manually without formatting, and the root partition as ext4. I will be using a mixture of spare drives just for SMB shares and I should be done.

**Reply by 1000001101000** [reply]
Nice!

I'm really only asking out of curiosity, trying to understand which devices people are using an what for. 

Have fun!

**Reply by enyafox** [reply]
Thank you, Its great what you've done here. I only got the NAS very cheap at a car boot sale. I got annoyed at the stock firmware and the first thing that popped up when searching was this and its been great using it.

**Reply by enyafox** [reply]
While we are here, how would you get the buzzer to beep through any of the python scripts?

**Reply by 1000001101000** [reply]
The example scripts in the repo include how to work pretty much everything.

I didn't create one specifically for the ls-xl but I think the one for the ts3000 should be basically the same. 

this plays a sound: test.cmd_sound(libmicon.BZ_MUSIC2)

I've made some documentation of the different commands here too:
https://buffalonas.miraheze.org/wiki/Terastation_Microcontroller_Interface

**Reply by enyafox** [reply]
Ah, thanks for that. I'm probably overthinking everything and missing lines and sections. 
Another thing, Have you tried any web gui's to manage or what would you recommend trying?

**Reply by 1000001101000** [reply]
I’ve seen people use webmin and omv. I don’t really recommend either since they’re surprisingly resource heavy and can mess things up if you aren’t careful.

**Reply by enyafox** [reply]
I have tried webmin and OMV and they make the NAS really slow. I just didn't know if you had any recommendations.

**Reply by 1000001101000** [reply]
I've never understood what Webmin and OMV do that requires so much memory/cpu. 

I played around with cockpit briefly a while back and it didn't seem quite as bad but it was mostly only good for displaying system information rather than configuring things. It looks like that project has progressed since then and there are some add-ons that might add value.

**Reply by enyafox** [reply]
Oh cool. I am currently installing cockpit and hopefully that would help

**Reply by enyafox** [reply]
Heya, just updating that I gave up on a web gui so I've just set up samba through ssh. I'm trying to have a script that reads the power button, for example for 5 seconds, and shuts down the NAS but I'm a little confused on how I would be able to that. I've had a look at the button test script but I'm finding it hard assigning the button codes to variables.

**Reply by 1000001101000** [reply]
oddly enough, I've never actually done that. 

off the top of my head I would write a short python script that checked the state of the button and then slept for 1 second (maybe even shorter). it would count the number of times it saw it as "pressed" and would call the desired command if the ocunt reached a certain number. you could start that with a systemd service basically the same as the fan one.

**Reply by enyafox** [reply]
Would you be able to write the script for me please? I haven't been able to figure something out. I also have another problem (Sorry for keep on going on lol) but when displaying text and scrolling through different things, the LCD screen seems to get darker and darker. Could it be a mistake i'm making?

**Reply by 1000001101000** [reply]
I can take a swing at it.

Could you post your current code?

**Reply by enyafox** [reply]
Unfortunately I've just cleared the file.

**Reply by enyafox** [reply]
I have taken the front panel off and apart and re-seated everything. The display seems to be working properly. I just need a way to use the buttons to do different things.

**Comment by 1000001101000** [comment]
I went ahead and wrote a script which shutsdown the system if you hold the power button:
https://github.com/1000001101000/Python_buffalo_libmicon/blob/master/button_daemon.py

if you put it in /usr/local/bin, mark it executable and run it as root you can test it out.

after you do that you can make it run at boot by creating the following systemd service.

/etc/systemd/system/button_daemon.service:
```
[Unit]
Description=Terastation Button Service

[Service]
Type=simple
ExecStart=/usr/local/bin/micon_scripts/button_daemon.py
Restart=always
RestartSec=5s

[Install]
WantedBy=multi-user.target
```
once you create the service file you enable it with:
`systemctl enable button_daemon.service`
`systemctl start button_daemon.service`

**Reply by enyafox** [reply]
Ah alright. Thanks so much. Will be very helpful. Especially finding something to have the function or the hidden reset button do something

**Reply by 1000001101000** [reply]
I also updated the other button script so that it outputs values that can be used with that daemon script. If you want to have multiple buttons do different things you would probably want to add some logic to the script for that rather than try to run multiple in parallel.

**Reply by enyafox** [reply]
I did edit your button test script on my side so it shows the numbers you have used in that button script

**Reply by enyafox** [reply]
Just compared mine to yours, they both are the same. What a coincidence

**Reply by 1000001101000** [reply]
Nice!

Feel free to send me a pull request if you come up with something that might be useful for others.

**Reply by enyafox** [reply]
and if your importing multiple libraries, you can use one line "import libmicon, os, time" unless your using "from … import …"

**Reply by enyafox** [reply]
I know its a little complex but here is something I'd find more easier to use https://github.com/enyafox/Python_buffalo_libmicon/blob/master/button_test.py

**Reply by enyafox** [reply]
It reads out a more accurate number that could be used

**Reply by 1000001101000** [reply]
I've updated the script to list the raw button register value (like yours) and kept the button breakdown. I also kept the logic that only prints when something changes (personal preference).

One thing to keep in mind is that the total value is the sum of all the buttons/switches including the switch on the back (where applicable). If you use that as your trigger value rather than breaking out the individual button you could end up with a scenario where the power button script only works if all the switches/buttons are in a certain state rather then just when the power button is depressed. 

It all depends on how complicated you want to get.

**Reply by enyafox** [reply]
The switch on the back of mine does not seem to show anything on the test script. Could be a problem with the switch. Ill hook up a multimeter to it and check if there's something wrong.

**Reply by 1000001101000** [reply]
It could be the switch isn't hooked up to the microcontroller for that model. I would have thought it would be but I'm not sure. 

There's a lot of variation on how the buttons/switches are connected across models which is part of why I left the button related code so generic.

**Reply by enyafox** [reply]
Ah alright. Ill have a look and follow the traces

**Reply by 1000001101000** [reply]
lol
For the arm devices they are generally connected either to the microcontroller or directly to a gpio pin.

On the Intel devices:
Power is connected as the ACPI power button
Display and info are connected to the IT8721 SuperIO chip
The switch on the back is hooked to an SoC gpio pin
I think there is also one hooked up to the microcontroller

they are....interesting... devices

**Reply by enyafox** [reply]
I have just contacted Buffalo, I'm actually surprised that their enquiries page still has my model listed, cause of its age, and Ill see what I can do to grab a schematic to share around and hopefully some source code so I can speed up boot process. Wish me luck lol

**Reply by 1000001101000** [reply]
They post source code here:
https://opensource.buffalo.jp/

 though they don't post their bootloader code for some reason (just generic uboot if anything at all).

**Reply by enyafox** [reply]
Ooh, I didn't know they had a page lol. Would you know how to use the switch if it was hooked up to a gpio pin?

**Reply by 1000001101000** [reply]
hmmm, I don't actually know for this device. 

I just booted mine up and took a look. I confirmed it doesn't show up when scanning the microcontroller buttons. I just tried scanning the gpio values while flipping the switch back and forth on mine but none of them changed. 

The support for the MV78100 SoC is less .... robust than others, it's possible that it's missing some of the gpio pins or possibly the button gpio was assigned to something else incorrectly. You might be able to figure out what the issue is if you can find a reference to the switch in Buffalo's source. 

I found a similar issue with the LS-QL which had a button and USB power regulator on "unsupported" pins.

**Reply by 1000001101000** [reply]
considering the header file for the SoC lists 49 GPIOs but the kernel is only exposing 32 I'm guessing that's exactly what's going on:
https://github.com/torvalds/linux/blob/master/arch/arm/mach-mv78xx0/mpp.h

**Reply by 1000001101000** [reply]
I took a look at the buffalo source and found this:
`  {BOARD_GPP_PWRAUTO_SW, 15},`

I then took a look at the relavant code in the Debian kernel and confirmed GPIO15 was configured for a non-existent SATA activity LED. I changed that to a generic GPIO and built a new kernel and installed it on my device. I'm now able to see GPIO15 change when I flip the switch back and forth. 

I'll work on getting that patch added to my kernel build process and then generate new kernels over the weekend.

**Reply by 1000001101000** [reply]
I've pushed new kernels with the GPIO15 fix. You should be able to install it with `apt-get update; apt-get upgrade linux-image-tsxl` and then rebooting.

Once you hae a suitable kernel installed the value of the switch can be read by 
1. exporting the pin with: `echo "15" > /sys/class/gpio/export`
2. set direction as 'in': `echo "in" > /sys/class/gpio/gpio15/direction`
2. checking the value of `/sys/class/gpio/gpio15/value`

**Reply by enyafox** [reply]
Thanks for that, it works great. I haven't had the time to have a look the past couple times. It would definitely use it to turn on or off the display backlight for night time as it does get bright lol

**Reply by 1000001101000** [reply]
Lol, seems like everytime someone asks about the TS-XL I end up writing a kernel patch

That sounds like an interesting use for the switch!

I think there is a newer way of accessing gpios from userspace which includes triggering that might be even more useful.. I do not recall the details as I've never used it though.

**Reply by enyafox** [reply]
I dont know. I've just used python to read the value through cat (I know it might not be the best solution but it works) and using if statements so its not always sending the lcd off or lcd on to the microcontroller. It just sits there until the switch changes

**Reply by enyafox** [reply]
So, I will use os.system() to export the pin on boot with the backlight script. Makes it all in one neat place I suppose lol

**Reply by 1000001101000** [reply]
Sounds like you have it under control!

**Reply by enyafox** [reply]
Yeah. Is there a way to spin down hard drives? It just makes the NAS quieter at night as it sits close to me lol. I've tried all the normal Debian stuff(I think) but they do seem to keep spinning

**Reply by enyafox** [reply]
I have managed to get them to sleep lol. When I have time, I'll have a look to see what's waking the noisy one up lol

**Reply by 1000001101000** [reply]
I haven't spent a ton of time on this topic with this device (or any other than my laptop really).

I believe the main issue with just spinning down the drives is that if your rootfs in on the drives it will spin them back up frequently to write out log entries from background tasks etc. There are ways around this but each comes with some challenges and some may not be possible for this specific device.

0. **Just replace your disks with SSDs** obviously, this may not be practical though running your rootfs on an old SSD and using the remaining bays for spinning data disks would work pretty well.
1. **Move the rootfs to a USB stick**. The device can't boot from usb but it is possible to have your /boot on the main drives but your roofts on a usb stick.
2. **Suspend the device to ram** I kind of doubt this SoC can do that with mainline linux and also doubt you could wake it if it did but I'm adding it to the list anyway.
3. **shutoff and wake via schedule with RTC** I also do not know if this is possible for this particular device using mainline linux but am somewhat more optimistic it is possible.
4. **shutoff and wake via WOL** I also do not know if this is possible for this particular device using mainline linux but am significantly more optimistic it is possible.

**Reply by enyafox** [reply]
So, they do seem to spin down when not in use. I do need to go online and have a look at a more quieter fan sometime. I have changed a few values but the NAS gets quite warm lol

**Reply by enyafox** [reply]
I do have the rootfs on an ssd in slot 1. I should really also look into a 3.5' converter for my SSD

**Reply by 1000001101000** [reply]
is the rootfs just on the SSD or is it raid1 across the drives like the stock firmware?

**Reply by enyafox** [reply]
The rootfs is just on the ssd and the rest are just normal ext4 drives

**Comment by 1000001101000** [comment]
@enyafox I recently discovered that the USB ports on this device are enabled via a set of GPIOs that weren't all getting activated at boot. I updated my kernel patch to enable them as part of the initialization process. While I was add it I also mapped GPIO15 as a button in the kernel.

Once you update the kernel you can start managing the rear switch like a normal button/key using triggerhappy etc.

**Reply by enyafox** [reply]
Ah  alright. I dont use the usb ports anyway but great work :)

While i'm here, would you know if I can watch the boot process as I am suddenly failing to boot?

**Reply by 1000001101000** [reply]
This particular model has a read-only serial console behind the door on the front, It's pretty easy to hook up the appropriate adapter and watch the boot process.

Supposedly with a bit of soldering you can even make it read/write, but I've never bothered with that:
https://web.archive.org/web/20180723055005/http://buffalo.nas-central.org/wiki/Terastation_III_Serial_console

**Reply by enyafox** [reply]
Alright thanks, do you know the baud rate that is used?

**Reply by enyafox** [reply]
And would you know what i could use to connect through serial?

**Reply by 1000001101000** [reply]
115200

I use minicom on linux, PuTTy should work on windows.

**Reply by enyafox** [reply]
Unless ive got the pins wrong, im just getting the normal scrambled characters

**Reply by 1000001101000** [reply]
scrambled characters usually means some issue like parity/speed/stop bits

**Comment by 1000001101000** [comment]
```
  +-----------------------------------------------------------------------+
    | A -    Serial Device      : /dev/ttyUSB0                              |
    | B - Lockfile Location     : /var/lock                                 |
    | C -   Callin Program      :                                           |
    | D -  Callout Program      :                                           |
    | E -    Bps/Par/Bits       : 115200 8N1                                |
    | F - Hardware Flow Control : Yes                                       |
    | G - Software Flow Control : No                                        |
    |                                                                       |
    |    Change which setting?                                              |
    +-----------------------------------------------------------------------+

```

**Reply by enyafox** [reply]
Alright, would you think that using a Raspberry pi would work through the serial pins on that?

**Reply by 1000001101000** [reply]
you keep saying pins, I don't understand quite what you are referring to

**Reply by 1000001101000** [reply]
I use something like this: 
https://www.amazon.com/Serial-Usb-Adapter/s?k=Serial+To+Usb+Adapter

**Reply by enyafox** [reply]
Still scrambled up characters

**Reply by 1000001101000** [reply]
I've done that with my TSXL as recently as last week, it should work with the right settings/etc. Can you provide more detail of what tools you are using?

**Reply by 1000001101000** [reply]
....maybe two weeks ago, whenever I was working on the kernel patch
